GH-35166: [C++][Compute] Increase precision of decimals in sum aggregates

khwilson opened the pull request:
### Rationale for this change

As documented in #35166, when Arrow performs a sum, product, or mean of an array of type `decimalX(P, S)`, it returns a scalar of type `decimalX(P, S)`. This is true even if the aggregate does not fit in the specified precision. For instance, a sum of two `decimal128(1, 0)`'s such as `1 + 9` is a `decimal128(2, 0)`. But (in Python):

```python
import pyarrow as pa
from decimal import Decimal

arr = pa.array([Decimal("1"), Decimal("9")], type=pa.decimal128(1, 0))
assert arr.sum().type == pa.decimal128(1, 0)
```

This is recognized in the rules for _binary_ addition and multiplication of decimals (see footnote 1 in [this section](https://arrow.apache.org/docs/cpp/compute.html#arithmetic-functions)), but this does not apply to array aggregates.

In #35166 I did a bit of research following a question from @westonpace , and it seems that there's no standard approach to this across DBMS's, but a common solution is to set the precision of the result of a sum to the maximum possible precision of the underlying type. That is, a sum of `decimal128(1, 0)`'s becomes a `decimal128(38, 0)`.

However, products and means differ further. For instance, in both instances, duckdb converts a decimal to a double, which makes sense as the precision of the product of an array of decimals would likely be huge, e.g., an array of size N with precision 2 decimals would have precision at least 2^N.

This PR implements the minimum possible change: replace the return types of the sum aggregate of `decimal128(P, S)` to `decimal128(38, S)`, `decimal256(P, S)` to `decimal256(76, S)`, `decimal32(P, S)` to `decimal32(9, S)`, and `decimal64(P, S)` to `decimal64(18, S)`.

### What changes are included in this PR?

- [x] Update C++ kernels to support the change
- [x] Update docs to reflect the change
- [x] Fix tests in the languages that depend on the C++ engine
- [x] Determine if there are other languages which do _not_ depend on the C++ engine which should also be updated

### Are these changes tested?

They are tested in the following implementations:

- [x] Python
- [x] C++

### Are there any user-facing changes?

Yes. This changes the return type of a scalar aggregate of decimals.

**This PR includes breaking changes to public APIs.**

Specifically, the return type of a scalar aggregate of decimals changes. This is unlikely to break downstream applications as the underlying data has not changed, but if an application relies on the (incorrect!) type information for some reason, it would break.
* GitHub Issue: #35166

khwilson commented:
@zeroshade I noticed in #43957 that you were adding in Decimal32/64 types, which I think will have the same problem that this PR addresses. I was curious if you might have interest in reviewing this PR?

zeroshade commented:
@khwilson Sure thing, i'll try to take a look at this in the next day or so

khwilson commented:
Hi @zeroshade just checking in! Thanks again for taking a look

mapleFU commented on cpp/src/arrow/compute/kernels/codegen_internal.h:
```diff
@@ -1380,6 +1383,8 @@ Status CastBinaryDecimalArgs(DecimalPromotion promotion, std::vector<TypeHolder>
 ARROW_EXPORT
 Status CastDecimalArgs(TypeHolder* begin, size_t count);
 
+Result<std::shared_ptr<DataType>> WidenDecimalToMaxPrecision(std::shared_ptr<DataType> type);
```
Should this being exported?

khwilson replied:
Done!

mapleFU commented on cpp/src/arrow/compute/kernels/codegen_internal.cc:
```diff
@@ -76,6 +76,26 @@ Result<TypeHolder> ListValuesType(KernelContext* ctx,
+                                                            base_type_->scale()));
+    }
+  } else {
+    return Status::TypeError(
+        "A call to MaxPrecisionDecimalType was made with a non-DecimalType");
```
would you mind point out the type here?

khwilson replied:
Simplified to just call the common `WidenDecimalToMaxPrecision` function which is just the identity on non-Decimal types

mapleFU replied:
I mean, TypeError and add the type here. Since we've supported decimal32 and decimal64

khwilson replied:
Done!

mapleFU commented on cpp/src/arrow/compute/kernels/codegen_internal.h:
```diff
@@ -55,6 +55,7 @@ using internal::BinaryBitBlockCounter;
 using internal::BitBlockCount;
 using internal::BitmapReader;
 using internal::checked_cast;
+using internal::checked_pointer_cast;
```
why in header?

khwilson replied:
I was following the pattern of the `internal::checked_cast` on the line above. But I'm not wedded to including it.

mapleFU reviewed (commented):
This method is interesting, however, before doing that, do you think a user-side "cast" is ok?
Like:

```
cast(origin to decimal(large-enough)) then avg
```

khwilson commented:
Thanks for the review!

By a user-side cast, do you mean that users should essentially do:

```sql
select avg(cast(blah as decimal(big-precision)))
```

instead of

```sql
select avg(blah)
```

or do you mean that this code should "inject" a cast on the "user" side?

If you mean putting the cast onto the user, then I would think you'd want to add an error if the answer can't fit into the default precision, but that seems like it would be more disruptive (and out of step with how other systems handle decimal aggregates).

If you mean "injecting" the cast on the user side, would that end up creating a copy of the array?

zeroshade commented:
@khwilson Hey sorry for the delay from me here, I've been traveling a lot lately for work and have been at ASF Community Over Code this week. I promise i'll get to this soon. In the meantime, you're in the very capable hands of @mapleFU

khwilson commented:
No problem! Hope your travels were fun!

mapleFU commented:
Generally this method is ok for me, but I'm not so familiar with the "common solutions" here. I'll dive into Presto/ClickHouse

mapleFU commented:
Would you mind making this Ready for review?

khwilson commented:
Sure!

khwilson commented:
@mapleFU I believe this is done now. Some notes on the diff:

* The hash aggregates had to be updated as well (missed them in the first pass)
* I've also added in Decimal32/64 support the basic aggregates (sum, product, mean, min/max, index). However, there's quite a lot of missing support for these types still in compute (most notably in casts)
* Docs are updated to reflect the change

And a note that quite a few tests are failing for what appears to be the same reason as #41390. Happy to address them if you'd like.

pitrou commented:
I'm lukewarm about the approach here. Silently casting to the max precision discards metadata about the input; it also risks producing errors further down the line (if e.g. the max precision is deemed too large for other operations). It also doesn't automatically eliminate any potential overflow, for example:
```python
>>> a = pa.array([789.3] * 20).cast(pa.decimal128(38, 35))
>>> a
<pyarrow.lib.Decimal128Array object at 0x7f0f103ca7a0>
[
  789.29999999999995452526491135358810440,
  789.29999999999995452526491135358810440,
  789.29999999999995452526491135358810440,
  789.29999999999995452526491135358810440,
  789.29999999999995452526491135358810440,
  789.29999999999995452526491135358810440,
  789.29999999999995452526491135358810440,
  789.29999999999995452526491135358810440,
  789.29999999999995452526491135358810440,
  789.29999999999995452526491135358810440,
  789.29999999999995452526491135358810440,
  789.29999999999995452526491135358810440,
  789.29999999999995452526491135358810440,
  789.29999999999995452526491135358810440,
  789.29999999999995452526491135358810440,
  789.29999999999995452526491135358810440,
  789.29999999999995452526491135358810440,
  789.29999999999995452526491135358810440,
  789.29999999999995452526491135358810440,
  789.29999999999995452526491135358810440
]
>>> pc.sum(a)
<pyarrow.Decimal128Scalar: Decimal('-1228.11834604692408266343214451664848480')>
```

We should instead check that the result of an aggregate fits into the resulting Decimal type, while overflows currently pass silently:
```python
>>> a = pa.array([123., 456., 789.]).cast(pa.decimal128(4, 1))
>>> a
<pyarrow.lib.Decimal128Array object at 0x7f0ed06261a0>
[
  123.0,
  456.0,
  789.0
]
>>> pc.sum(a)
<pyarrow.Decimal128Scalar: Decimal('1368.0')>
>>> pc.sum(a).validate(full=True)
Traceback (most recent call last):
  ...
ArrowInvalid: Decimal value 13680 does not fit in precision of decimal128(4, 1)
```

khwilson commented:
Two problems with just validating afterward: First, I'd expect in reasonable cases for the validation to fail. A sum of 1m decimals of approximately the same size you'd expect to have 6 more digits of precision. I assume this is why all the DBMSs I looked at increase the precision by default.

Second, just checking for overflow doesn't solve the underlying problem. Consider:

```python
a = pa.array([789.3] * 18).cast(pa.decimal128(38, 35))
print(pc.sum(a))
pc.sum(a).validate(full=True)  # passes
```

In duckdb, they implement an intermediate check to make sure that there's not an internal overflow:

```python
tab = pa.Table.from_pydict({"a": a})
duckdb.query("select sum(a) from tab")
# Traceback (most recent call last):
#   File "<stdin>", line 1, in <module>
# duckdb.duckdb.OutOfRangeException: Out of Range Error: Overflow in HUGEINT addition: 
# 157859999999999990905052982270717620880 + 78929999999999995452526491135358810440
```

Notably, this lack of overflow checking also applies to integer sums in arrow:

```python
>>> pa.array([9223372036854775800] * 2, type=pa.int64())
<pyarrow.lib.Int64Array object at 0x10c1d8b80>
[
  9223372036854775800,
  9223372036854775800
]
>>> pc.sum(pa.array([9223372036854775800] * 2, type=pa.int64()))
<pyarrow.Int64Scalar: -16>
>>> pc.sum(pa.array([9223372036854775800] * 2, type=pa.int64())).validate(full=True)
```

pitrou commented:
> Two problems with just validating afterward: First, I'd expect in reasonable cases for the validation to fail. A sum of 1m decimals of approximately the same size you'd expect to have 6 more digits of precision.

It depends obviously if all decimals are of the same sign, and what their actual magnitude is. 

> Second, just checking for overflow doesn't solve the underlying problem.

In the example above, I used a `validate` call simply to show that the result was indeed erroneous. I didn't mean we should actually call validation afterwards. We should instead check for overflow at each individual aggregation step (for each add or multiply, for example). This is required _even_ if we were to bump the result's precision to the max.

pitrou commented:
> Notably, this lack of overflow checking also applies to integer sums in arrow:

Yes, and there's already a bug open for it: https://github.com/apache/arrow/issues/37090

khwilson commented:
Nice! I'm excited for the checked variants of sum and product!

With the integer overflow example, I only meant to point out that the compute module currently allows overflows, so I think it would be unexpected for `sum` to complain about an overflow _only_ if the underlying type was a decimal. But if the goal with #37090 is to _replace_ `sum` with a checked version, then the solution of erroring makes a lot of sense, and I'd be happy to implement it when #37536 gets merged. :-)

Still, I do think that users would find it unexpected to get an error if the sum fit in the underlying storage since this is how all the databases I've used (and the four I surveyed in #35166) have operated.

khwilson commented:
Hi @pitrou sorry for dropping the ball on this a bit earlier. Are you still interested in merging this at some point? I'm happy to explore some alternative options for how to do the decimal expansion, or alternatively am happy to push #37536 forward

khwilson commented:
Hey @pitrou i was thinking about this PR again recently. Would you still be interested in merging if I cleaned it up a bit?

pitrou commented:
@khwilson I'm still not sure this is actually desirable (@zanmato1984 what do you think?).

What I think is definitely desirable is to check for overflow when doing sum/mean/product of decimals, regardless of which data type is selected for output.

khwilson commented:
Cool. I’m happy to do some more clean up or do a little work on creating the checked versions of the aggregates to push this along

zanmato1984 commented:
> Cool. I’m happy to do some more clean up or do a little work on creating the checked versions of the aggregates to push this along

Thank you for the contribution and passion. How about we keep this PR focused solely on the precision promotion? And the potential checked version can be done in another PR.

khwilson commented:
That makes sense. From a sequencing stand point, should the checked versions come first or should we get this through first?

zanmato1984 commented:
I think they are well independent so feel free to start anytime. And thank you for willing to take this!

zanmato1984 commented on cpp/src/arrow/compute/kernels/hash_aggregate.cc:
```diff
@@ -564,7 +564,7 @@ struct GroupedReducingAggregator : public GroupedAggregator {
   static enable_if_decimal<T, std::shared_ptr<DataType>> GetOutType(
       const std::shared_ptr<DataType>& in_type) {
-    return in_type;
+    return WidenDecimalToMaxPrecision(in_type).ValueOrDie();
```
I don't think we should widen the output type for all the subclasses of `GroupedReducingAggregator`. Promoting to max precision is reasonable for `sum` but is not necessarily the case for others.

khwilson replied:
Yeah, different databases do different things. E.g., for a mean, duckdb converts to a double (which is also what the array aggregate does for non-decimal numeric types in arrow). Do you have a preference for what the type for the other aggregates _should_ be?

zanmato1984 commented on cpp/src/arrow/compute/kernels/aggregate_basic.cc:
```diff
@@ -365,11 +365,16 @@ struct ProductImpl : public ScalarAggregator {
   }
 
   Status Finalize(KernelContext*, Datum* out) override {
+    std::shared_ptr<DataType> out_type_ = this->out_type;
+    if (is_decimal(this->out_type->id())) {
+      ARROW_ASSIGN_OR_RAISE(out_type_, WidenDecimalToMaxPrecision(this->out_type));
+    }
+
```
This is hijacking the output type at the execution time. However we should be consistent between the function signature (the return type determined at the function resolution) and the value actually being returned.

You should probably change the return type somewhere in https://github.com/apache/arrow/blob/067fd2a2c6e54d33b9ae8a3324f59bebe960d485/cpp/src/arrow/compute/kernels/aggregate_basic.cc#L1051-L1054 .
(BTW, what you did for hash aggregates is correct - widening the precision to the actual output type rather than hijacking the return value at the last minute).

khwilson replied:
The link you have is to a different commit, I believe. If you look at those lines in this PR they do change the type at resolution time.

khwilson replied:
I can't remember exactly why I had to do this. I believe it's related to how the `Visit` function gets called for `SumLikeImpl`. Once the rebase is done I can take a look!

zanmato1984 replied:
> The link you have is to a different commit, I believe. If you look at those lines in this PR they do change the type at resolution time.

Ah, I see changing the output type in resolution time. Thanks for clarifying.

Then the question is why changing the output type again during the execution, which I think you are figuring out :)

zanmato1984 reviewed (changes requested):
Other than the inline comments, some other things:
1. Please rebase the PR to the latest main.
2. I think we should exclude the addition of `Decimal32/64` support from this PR, keeping it focused on the precision promotion, preferably for `sum` and optionally `mean`.

khwilson commented:
OK, I think I've correctly rebased (letting all the tests run now to figure out).

I can remove the Decimal32/64 support once I check if I've rebased correctly.

Will also look at the inline comments once the rebase is completed.

khwilson commented:
OK @zanmato1984 it took a bit given the way that everything gets set. Updates:

* Removed all the Decimal32/64 logic (except in `WidenDecimalToMaxPrecision`
* Introduced a `PromoteDecimal` template variable to determine whether a `SumLike` should actually promote the decimal
* Reverted the Mean and Product promotions as we're not agreed on how to handle those.

As for the mean and product, there's a few actions we could take:

* Status quo: Input type is the same as output type
* Promote both to maximum precision
* Promote only _product_ to max precision: `abs(mean) <= max(abs(max), abs(min))` so not really _necessary_ for mean
* Promote only _product_ to a double: this is what, e.g., duckdb does

I think the best course of action is actually the fourth choice: Promote only product to double and leave mean the same as the input type. Thoughts?

pitrou commented:
> As for the mean and product, there's a few actions we could take:
>     * Status quo: Input type is the same as output type
>     * Promote both to maximum precision
>     * Promote only _product_ to max precision: `abs(mean) <= max(abs(max), abs(min))` so not really _necessary_ for mean
>     * Promote only _product_ to a double: this is what, e.g., duckdb does

"Promoting" to double is a demotion: it loses a lot of potential precision. The mean should certainly stay the same as the input type (and care should be taken to avoid overflows during intermediate computations, if at all possible).

As for the product, it would also be better to not lose too much precision, and therefore stay in the decimal domain. But it sounds less common to compute the product of a bunch of decimals, anyway.

khwilson commented:
If I understand the code correctly, what happens for the mean for decimals is that it first computes the sum by using the underlying `+` of the Decimal type, then computes the count of values as a long, and does an integer division of the two. As seen in this commit, that `+` operator essentially ignores precision. I think this is a reasonable implementation except:

* There should be a checked variant;
* Potentially, you'd want to add some extra scale, in line with the binary division operation.

For comparison, I checked how a few other DBMS's handle the mean:
* postgresql: Drops all precision and scale information in mean
* duckdb: Turns into a double
* mariadb: Appears to just add 4 to scale and precision

For product, it's not very common for databases to provide a product aggregator of any kind. I quickly looked at ANSI SQL, postgres, mysql, and clickhouse, none of which provide a product. The "standard" product implementation for such aggregates is `(2 * sum(col % 2) - 1) * exp(sum(ln(abs(col)))`, but that requires transforming into a double. Moreover, basically any column with at least 100 values will overflow a decimal's precision.

So I'd propose for product either demoting to double (will give a semantically correct value but lose precision) or maxing out the width of the decimal type. Of course, there should also be a checked variant (but I'm deferring that until after this commit is complete).

pitrou commented:
> If I understand the code correctly, what happens for the mean for decimals is that it first computes the sum by using the underlying `+` of the Decimal type, then computes the count of values as a long, and does an integer division of the two. As seen in this commit, that `+` operator essentially ignores precision. I think this is a reasonable implementation except:
>     * There should be a checked variant;
>     * Potentially, you'd want to add some extra scale, in line with the binary division operation.

Ideally, the scale should be "floating" just as in floating-point arithmetic, depending on the current running sum (the running sum can be very large if all data is positive, or very small if the data is centered around zero). It is then normalized to the original scale at the end. But of course that makes the algorithm more involved.

(that would also eliminate the need for a checked variant?)

> So I'd propose for product either demoting to double (will give a semantically correct value but lose precision) or maxing out the width of the decimal type. Of course, there should also be a checked variant (but I'm deferring that until after this commit is complete).

Either is fine to me.

khwilson commented:
> Ideally, the scale should be "floating" just as in floating-point arithmetic, depending on the current running sum (the running sum can be very large if all data is positive, or very small if the data is centered around zero). It is then normalized to the original scale at the end. But of course that makes the algorithm more involved.

Yeah, this is quite complicated and essentially means you need to implement the floating point addition algorithm.

> Either is fine to me.

Cool. @zanmato1984 do you have an opinion?

zanmato1984 commented on cpp/src/arrow/compute/kernels/aggregate_basic.inc.cc:
```diff
@@ -187,10 +187,20 @@ struct SumLikeInit {
     return Status::OK();
   }
 
+  /* Decimal types
+   *
+   * By default, we widen the decimal to max precision for SumLikes
+   * However, this may not be the desired behaviour (see, e.g., MeanKernelInit) */
```
We tend to use `///` for docstring.

khwilson replied:
Done!

zanmato1984 commented on cpp/src/arrow/compute/kernels/aggregate_basic.inc.cc:
```diff
@@ -187,10 +187,20 @@ struct SumLikeInit {
-    state.reset(new KernelClass<Type>(type, options));
-    return Status::OK();
+    if constexpr (PromoteDecimal) {
+      auto ty = WidenDecimalToMaxPrecision(type).ValueOrDie();
```
```suggestion
      ARROW_ASSIGN_OR_RAISE(auto ty, WidenDecimalToMaxPrecision(type));
```

khwilson replied:
Done.

zanmato1984 commented on cpp/src/arrow/compute/kernels/hash_aggregate_numeric.cc:
```diff
@@ -162,7 +164,11 @@ struct GroupedReducingAggregator : public GroupedAggregator {
       const std::shared_ptr<DataType>& in_type) {
-    return in_type;
+    if constexpr (PromoteDecimal) {
+      return WidenDecimalToMaxPrecision(in_type).ValueOrDie();
```
```suggestion
      return WidenDecimalToMaxPrecision(in_type);
```

zanmato1984 replied:
You might want to keep the error propagation.

khwilson replied:
See below.

zanmato1984 commented on cpp/src/arrow/compute/kernels/hash_aggregate_numeric.cc:
```diff
@@ -162,7 +164,11 @@ struct GroupedReducingAggregator : public GroupedAggregator {
   template <typename T = Type>
   static enable_if_decimal<T, std::shared_ptr<DataType>> GetOutType(
```
```suggestion
  static enable_if_decimal<T, Result<std::shared_ptr<DataType>>> GetOutType(
```

khwilson replied:
Makes sense. Changed.

zanmato1984 reviewed (changes requested):
Some initial suggestions. More coming.

khwilson commented:
I agree that it doesn't need to be answered in this PR. Will tackle in a future PR. :-)

I also made the changes requested.

zanmato1984 commented:
Sorry @khwilson for being late. I wasn't able to make my time for this in last few weeks. I will look into it again in this or next week. Thanks for the patient, appreciate it!

khwilson commented:
No problem! Thanks for the reviews!

zanmato1984 commented on cpp/src/arrow/compute/kernels/hash_aggregate_numeric.cc:
```diff
@@ -42,7 +42,8 @@ namespace {
 template <typename Type, typename Impl,
-          typename AccumulateType = typename FindAccumulatorType<Type>::Type>
+          typename AccumulateType = typename FindAccumulatorType<Type>::Type,
+          bool PromoteDecimal = false>
```
Ditto.

zanmato1984 commented on cpp/src/arrow/compute/kernels/aggregate_basic.inc.cc:
```diff
@@ -187,10 +187,18 @@ struct SumLikeInit {
     return Status::OK();
   }
 
+  /// By default, we widen the decimal to max precision for SumLikes
+  /// However, this may not be the desired behaviour (see, e.g., MeanKernelInit)
```
These comments might be unnecessary now?

zanmato1984 commented on docs/source/cpp/compute.rst:
```diff
@@ -264,7 +264,9 @@ the input to a single output value.
 
 * \(5) For decimal inputs, the resulting decimal will have the same
-  precision and scale. The result is rounded away from zero.
+  scale and the maximum precision for its width. For instance, an array
```
This is for `mean` right? And we didn't promote for `mean` in this PR.

khwilson replied:
Good catch. Missed on the revision.

zanmato1984 reviewed (changes requested):
Haven't looked at the tests yet. These are my comments so far.

I've also pushed a commit to simplify `WidenDecimalToMaxPrecision`.

khwilson commented:
Thanks, Rossi! My turn to apologize on tardiness. Will address this weekend.

On Thu, Jun 5, 2025 at 1:41 PM Rossi Sun ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In cpp/src/arrow/compute/kernels/codegen_internal.cc
> <https://github.com/apache/arrow/pull/44184#discussion_r2129530668>:
>
> > @@ -528,6 +533,28 @@ Status CastDecimalArgs(TypeHolder* begin, size_t count) {
>    return Status::OK();
>  }
>
> +Result<std::shared_ptr<DataType>> WidenDecimalToMaxPrecision(
>
> Might want to add some test cases for this function in
> codegen_internal_test.cc.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/arrow/pull/44184#pullrequestreview-2901370679>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AALU5EX7XE4NS3F6BSYIL5L3CB6MJAVCNFSM6AAAAABOVDOQEWVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDSMBRGM3TANRXHE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

khwilson commented on cpp/src/arrow/compute/kernels/codegen_internal.cc:
```diff
@@ -528,6 +533,28 @@ Status CastDecimalArgs(TypeHolder* begin, size_t count) {
   return Status::OK();
 }
 
+Result<std::shared_ptr<DataType>> WidenDecimalToMaxPrecision(
```
Done.

khwilson commented:
OK, @zanmato1984 added the tests and updated docs.

For the SFINAE things, a couple considerations:
* Probably isn't super critical for performance, but...
* In the grouped aggregators, `GetOutType` is `static` so creating class-level (e.g., `GroupedMeanImpl`) runtime differences would likely require refactoring `GetOutType`.

Given the second bullet, I'd keep it as SFINAE for consistency between the two implementations.

zanmato1984 commented:
I think it would be less impactful in terms of design consequence if the changes are all in `.cc` file. So we should be good.

zanmato1984 commented:
Oh @khwilson one more thing, please update the PR description based on the latest code changes. Thanks.

khwilson commented:
Done

pitrou commented on cpp/src/arrow/compute/kernels/hash_aggregate_numeric.cc:
```diff
@@ -259,8 +271,11 @@ struct GroupedReducingFactory {
-  using Base = GroupedReducingAggregator<Type, GroupedSumImpl<Type>>;
+struct GroupedSumImpl
+    : public GroupedReducingAggregator<Type, GroupedSumImpl<Type>,
+                                       typename FindAccumulatorType<Type>::Type> {
```
I'm curious, why was this change necessary?

pitrou commented on docs/source/cpp/compute.rst:
```diff
@@ -171,8 +171,8 @@ To avoid exhaustively listing supported types, the tables below use a number
 * "Numeric": Integer types (Int8, etc.) and Floating-point types (Float32,
-  Float64, sometimes Float16).  Some functions also accept Decimal128 and
-  Decimal256 input.
+  Float64, sometimes Float16).  Some functions also accept
+  Decimal128/256 input.
```
More and more functions now also accept decimal32/decimal64.
```suggestion
  Float64, sometimes Float16).  Some functions also accept Decimal input.
```

khwilson replied:
The initial change to this line was to say that "some functions also accept Decimal32/64/128/256 input" but that change was rolled back to make this PR focus on Decimal128/256. Hence the change here. But this seems fine.

pitrou commented on cpp/src/arrow/compute/kernels/aggregate_test.cc:
```diff
@@ -495,51 +495,67 @@ TEST_F(TestSumKernelRoundOff, Basics) {
+                                                      decimal256(76, 2)};
+
+  for (size_t i = 0; i < init_types.size(); ++i) {
+    auto& ty = init_types[i];
+    auto& out_ty = out_types[i];
```
Nit: make those const
```suggestion
    const auto& ty = init_types[i];
    const auto& out_ty = out_types[i];
```

pitrou commented on cpp/src/arrow/compute/kernels/aggregate_test.cc:
```diff
@@ -495,51 +495,67 @@ TEST_F(TestSumKernelRoundOff, Basics) {
+                ResultWith(ScalarFromJSON(out_ty, R"(null)")));
     EXPECT_THAT(Sum(ScalarFromJSON(ty, R"(null)"), options),
-                ResultWith(ScalarFromJSON(ty, R"("0.00")")));
+                ResultWith(ScalarFromJSON(out_ty, R"("0.00")")));
```
Ok, but can we also test the effects of the added precision? i.e. test a case that would fail with the previous behavior of not increasing precision.

pitrou replied:
There may be a misunderstanding, but I was talking about the situation where the Sum's result would overflow the input type (be it decimal128 or decimal256).

pitrou commented on python/pyarrow/tests/test_compute.py:
```diff
@@ -355,10 +355,55 @@ def test_sum_array(arrow_type):
 
     arr = pa.array([], type=arrow_type)
     assert arr.sum().as_py() is None  # noqa: E711
+    assert pc.sum(arr).as_py() is None  # noqa: E711
```
You don't have to test both `arr.sum()` and `pc.sum(arr)` everytime since the former is just calling the latter. This will make the tests more readable and easier to maintain.

khwilson replied:
Sure, but as it was, this test tested both `arr.sum()` and `pc.sum(arr)` for each of its subcases _except_ this one. The correct change would be to remove all those checks, which seemed like a larger change than making it consistent.

khwilson replied:
Hobgoblins etc. ☺️ happy to remove these double checks as long as it’s the correct style moving forward!

khwilson replied:
Done.

khwilson commented:
OK, all in all:
* @pitrou removed one of `pc.sum` and `arr.sum` in the python test
* Added an example of an overflow in the sum tests
* Made a small modification to the docs to clarify that current changes apply only to sums.

wirable23 commented:
Would this change be part of arrow 21?

zanmato1984 commented:
@pitrou did you get a chance to look at this again?

pitrou commented on cpp/src/arrow/compute/kernels/hash_aggregate_numeric.cc:
```diff
@@ -154,17 +154,29 @@ struct GroupedReducingAggregator : public GroupedAggregator {
+    }
   }
 
+  /// If this returns true, then the aggregator will promote a decimal
```
Kind of a nit, but it might be better to avoid docstring notation (`///`) in `.cc` files because those are not APIs for which we want to expose a documentation.

pitrou reviewed (approved):
LGTM on the principle, I posted a couple comments that you might want to act upon. I'll let @zanmato1984 make the final call.

zanmato1984 commented on docs/source/cpp/compute.rst:
```diff
@@ -276,7 +275,9 @@ the input to a single output value.
 
 * \(5) For decimal inputs, the resulting decimal will have the same
-  precision and scale. The result is rounded away from zero.
+  scale and the precision for its width. For instance, an array
+  of ``decimal128(3, 2)`` will return a ``decimal128(3, 2)`` scalar.
+  The result is rounded away from zero.
```
```suggestion
  precision and scale. The result is rounded away from zero.
```

zanmato1984 commented on docs/source/cpp/compute.rst:
```diff
@@ -276,7 +275,9 @@ the input to a single output value.
 
 * \(5) For decimal inputs, the resulting decimal will have the same
-  precision and scale. The result is rounded away from zero.
+  scale and the precision for its width. For instance, an array
```
Now fully reverted.

pitrou commented on docs/source/cpp/compute.rst:
```diff
@@ -294,8 +295,10 @@ the input to a single output value.
+* \(9) Output is Int64, UInt64, Float64, or Decimal128/256, depending
+  on the input type. For sums of decimals, the precision is increased to the maximum
+  precision for the type's width. For instance, an array of
+  ``decimal128(3, 2)`` will return a ``decimal128(38, 2)`` scalar.
```
Ahah, no, I misread 3 for 38. Sorry.

zanmato1984 commented:
I pushed several commits to address the latest comments from @pitrou and me. I'll merge after CI is green.

@amoeba do you think we can include this in the 21.0.0 release?

zanmato1984 commented:
CI is good, merging.

Thanks a lot @khwilson for contributing this!

@wirable23 Please wait for our release manager's response about into 21.0.0. Thanks.
